src/libraries/System.Text.Json/Common/JsonHelpers.cs

```diff
@@ -22,6 +23,42 @@ internal static partial class JsonHelpers
             return false;
 #else
             return dictionary.TryAdd(key, value);
+#endif
+        }
+
+        /// <summary>
+        /// Provides an in-place, stable sorting implementation for List.
+        /// </summary>
+        internal static void StableSortByKey<T, TKey>(this List<T> items, Func<T, TKey> keySelector)
```

Comment:
cc @GrabYourPitchforks @tannergooding who might be interested in this implementation.